File: bookmarker-java/src/main/java/com/sivalabs/bookmarker/web/advice/GlobalExceptionHandler.java
package com.sivalabs.bookmarker.web.advice;

import lombok.extern.slf4j.Slf4j;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.servlet.ModelAndView;

@ControllerAdvice
@Slf4j
public class GlobalExceptionHandler  {

    @ExceptionHandler(AccessDeniedException.class)
    public String handleAccessDeniedException(AccessDeniedException e) {
        log.error(e.getMessage(), e);
        return "403";
    }

    @ExceptionHandler(Exception.class)
    public ModelAndView handleException(Exception e) {
        log.error(e.getMessage(), e);
        ModelAndView mav = new ModelAndView();
        mav.addObject("exception", e);
        mav.addObject("message", e.getMessage());
        mav.setViewName("error");

        return mav;
    }
}

File: src/main/kotlin/com/sivalabs/bookmarker/controller/GlobalExceptionHandler.kt
package com.sivalabs.bookmarker.controller

import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@RestControllerAdvice
class GlobalExceptionHandler : ResponseEntityExceptionHandler() {

    @ExceptionHandler(value = [EmptyResultDataAccessException::class])
    internal fun handleEmptyResultDataAccessException(exception: Exception): ResponseEntity<*> {
        return ResponseEntity(exception, HttpStatus.NOT_FOUND)
    }
}
